standards/JWT-SVID.md

```diff
@@ -0,0 +1,112 @@
+# The JWT SPIFFE Verifiable Identity Document
+
+## Status of this Memo
+This document specifies an experimental identity document standard for the internet community, and requests discussion and suggestions for improvements. It is a work in progress. Distribution of this document is unlimited.
+
+## Table of Contents
+
+1\. [Introduction](#1-introduction)  
+2\. [JOSE HEader](#2-jose-header)  
+2.1. [Algorithm](#21-algorithm)  
+2.2. [Key ID](#22-key-id)  
+2.3. [Type](#23-type)  
+3\. [JWT Claims](#3-jwt-claims)  
+3.1. [Subject](#31-subject)  
+3.2. [Audience](#32-audience)  
+3.3. [Expiration Time](#33-expiration-time)  
+4\. [Token Signing and Validation](#4-token-signing-and-validation)  
+5\. [Token Transmission](#5-token-transmission)  
+5.1. [Serialization](#51-serialization)  
+5.2. [HTTP](#52-http)  
+5.3. [gRPC](#53-grpc)  
+6\. [Security Considerations](#6-security-considerations)  
+6.1. [Replay Protection](#61-replay-protection)  
+6.2. [Audience](#62-audience)  
+6.3. [Transport Security](#63-transport-security)  
+Appendix A. [Validation Reference](#appendix-a-validation-reference)  
+
+## 1. Introduction
+JWT-SVID is the first token-based SVID in the SPIFFE specification set. Aimed at providing immediate value in solving difficulties associated with asserting identity across Layer 7 boundaries, compatibility with existing applications and libraries is a core requirement.
+
+JWT-SVIDs are standard JWT tokens with a handful of restrictions applied. JWTs have historically proven difficult to get right, gaining a reputation in the security community as a technology which is likely to introduce vulnerabilities in its deployments and implementations. JWT-SVID takes steps to mitigate these problems as much as possible without breaking compatibility with existing applications and libraries.
+
+## 2. JOSE Header
+Historically, complexity introduced by the cryptographic agility of the JOSE header has led to a series of vulnerabilities in popular JWT implementations. To avoid such pitfalls, this specification greatly restricts the allowances originally afforded. This section describes the permitted registered headers, as well as their values. Any header not described here, registered or private, MUST NOT be included in the JWT-SVID JOSE Header.
+
+Only JWS is supported, and all header values MUST reside in a JWS Protected Header.
+
+### 2.1. Algorithm
+The `alg` header MUST be set to `RS256`.
+
+Most vulnerabilities in JWT come from mixing and matching of cryptographic techniques, so JWT-SVID defines only a single technique as valid. By taking this approach, we aim to sidestep the pitfalls associated with cryptographic agility. In the event that the above-defined technique needs to be changed, it will be done by releasing a new version of JWT-SVID.
```

Comment:
Request for public comment on this point. Will imposing this restriction do more harm than good by hindering adoption in projects which support configurable key types and sizes?

Reply:
Ported comment by @gcallaghan from draft spec:
```OIDC recommends collision resistant names for custom claims https://openid.net/specs/openid-connect-core-1_0.html#AdditionalClaims, in practice implementers accomplish this by using a URI namespace https://auth0.com/docs/api-auth/tutorials/adoption/scope-custom-claims#custom-claims. This might be a worthwhile addition to the spec. Also, it might be nice to have guidance for some .well_known url for a consumer to advertise supported claims.```

Reply:
@gcallaghan

Reply:
I think the easies way to help validators conform is to provide a schema for them to match. Since JWT is a JSON document it makes sense to  use a JSON Schema. An example of one that would take this spec and the comments into consideration would be:

```yaml
$id: "https://spiffe.com/jwt-svd.json"
$schema: "http://json-schema.org/draft-07/schema#"
description: "SPIFFE schema to validate JWT SVD"
type: "object"
properties:
  header:
    type: "object"
    properties:
      alg:
        type: "string"
        description: |
          The alg (algorithm) Header Parameter identifies the cryptographic
          algorithm used to secure the JWS.  The JWS Signature value is not
          valid if the alg value does not represent a supported algorithm or
          if there is not a key for use with that algorithm associated with the
          party that digitally signed or MACed the content. The alg value is a case-
          sensitive ASCII string containing a StringOrURI value. This Header
          Parameter MUST be present and MUST be understood and processed by
          implementations.
          see https://tools.ietf.org/html/rfc7515#section-4.1.1
        enum:
          - "RS256"
          - "RS384"
          - "RS512"
          - "ES256"
          - "ES384"
          - "ES512"
      kid:
        type: "string"
        description: |
          The kid (key ID) Header Parameter is a hint indicating which key
          was used to secure the JWS.  This parameter allows originators to
          explicitly signal a change of key to recipients.  The structure of
          the kid value is unspecified.  Its value MUST be a case-sensitive
          string.  Use of this Header Parameter is OPTIONAL.
          see https://tools.ietf.org/html/rfc7515#section-4.1.4
    required:
      - "alg"
  payload:
    type: "object"
    properties:
      aud:
        type: "string"
        description: |
          A JSON string value, with the additional requirement that while
          arbitrary string values MAY be used, any value containing a ":"
          character MUST be a URI [RFC3986].  StringOrURI values are
          compared as case-sensitive strings with no transformations or
          canonicalizations applied.
          see https://tools.ietf.org/html/rfc7519#section-4.1.3
      exp:
        type: "integer"
        description: |
          The "exp" (expiration time) claim identifies the expiration time on
          or after which the JWT MUST NOT be accepted for processing.  The
          processing of the "exp" claim requires that the current date/time
          MUST be before the expiration date/time listed in the "exp" claim.
          Implementers MAY provide for some small leeway, usually no more than
          a few minutes, to account for clock skew.  Its value MUST be a number
          containing a NumericDate value.  Use of this claim is OPTIONAL.
          see https://tools.ietf.org/html/rfc7519#section-4.1.4
    required:
      - "aud"
      - "exp"
```

Reply:
Requiring RSA to the exclusion of elliptic curve algorithms seems like a pretty big negative to me.  While alg none is obviously dangerous, Elliptic curves offer a number of advantages over RSA, most notably in size, which can be nice in contexts when you're counting packets over the wire.

fwiw, github.com/square/go-jose simply does not implement alg: none.